### PR TITLE
[MERGE] web,*: introduce Layout component

### PR DESCRIPTION
**Pull request by Arcasias**
This PR introduces a new component: the layout component.
Its purpose is to make abstraction of most generic components and
classes when writing a view template or an action.

To make use of this component, we also needed a new system to pass slot
from the parent of the layout to the control panel. This was done in a
hacky way and is meant to be replaced as soon as Owl provides a built-in
way to pass slots.

The control panel has also been slightly tweaked to default its props
from an environment's newly introduced key: the 'config'. This has been
done to reduce the amount of props passed to the layout and search
subcomponents.

### CODE REVIEW

File: addons/web/static/src/views/view_layout.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+const { Component } = owl;
+
+export class ViewLayout extends Component {
+    setup() {
+        this.emitSlots = useSlotsEmitter({ exclude: ["default"] });
+    }
+
+    get ControlPanel() {
+        return this.props.ViewClass.components.ControlPanel || ControlPanel;
```
Comment (Arcasias) [inline addons/web/static/src/views/view_layout.js:16]: Put this in props (also search panel)

File: addons/web/static/src/views/slots_transmitter.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ *          <t t-slot="a" />
+ *      </div>
+ *      <t t-slot="default" />
+ *  </div>
+ *
+ * @param {Object} [params={}]
+ * @param {string[]} [params.exclude=[]]
+ * @returns {(props: Object = {}) => Object}
+ */
+export const useSlotsEmitter = ({ exclude = [] } = {}) => {
```
Comment (Arcasias) [inline addons/web/static/src/views/slots_transmitter.js:82]: doNotUseSlotEmitter

File: addons/web/static/src/views/view_layout.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/** @odoo-module **/
+
+import { ControlPanel } from "@web/search/control_panel/control_panel";
+import { SearchPanel } from "@web/search/search_panel/search_panel";
+import { Model } from "@web/views/helpers/model";
+import { useSlotsEmitter } from "@web/views/slots_transmitter";
+
+const { Component } = owl;
+
+export class ViewLayout extends Component {
```
Comment (Arcasias) [inline addons/web/static/src/views/view_layout.js:10]: ViewLayout => (Main)?Layout

File: addons/web/static/src/search/search_model.js [hunk 4]
```diff
@@ -1393,6 +1397,7 @@ export class SearchModel extends EventBus {
         return {
             controlPanel: "controlPanel" in display ? display.controlPanel : {},
             searchPanel:
+                this.sections.size &&
```
Comment (Polymorphe57) [inline addons/web/static/src/search/search_model.js:1400]: seems that it should be part of another commit

File: addons/web/static/src/views/view.js [hunk 5]
```diff
@@ -124,10 +125,9 @@ export class View extends Component {
         const type = ViewClass.type;
 
         // determine views for which descriptions should be obtained
-        let { views, viewId, searchViewId } = this.props;
-
-        views = JSON.parse(JSON.stringify(views));
+        let { viewId, searchViewId } = this.props;
 
+        const views = deepCopy(this.props.views || this.props.action.views);
```
Comment (Polymorphe57) [inline addons/web/static/src/views/view.js:130]: smth is wrong here: we should not have this alternative

Comment (Polymorphe57) [reply addons/web/static/src/views/view.js:130]: did you let this.props.views to make tests pass?

File: addons/web/static/src/search/search_model.js (hunk 4 above)
Comment (aab-odoo) [reply addons/web/static/src/search/search_model.js:1400]: shouldn't this be tested?

File: addons/web/static/src/search/control_panel/control_panel.xml [hunk 6]
```diff
@@ -41,9 +41,9 @@
 
             <div class="o_cp_pager"/>
 
-            <t t-if="props.viewSwitcherEntries and props.viewSwitcherEntries.length > 1">
+            <t t-if="viewSwitcherEntries.length">
```
Comment (aab-odoo) [inline addons/web/static/src/search/control_panel/control_panel.xml:44]: does it always exist? Seems weird for client actions

File: addons/web/static/src/search/with_search/with_search.js [hunk 7]
```diff
@@ -116,6 +106,20 @@ WithSearch.props = {
         },
         optional: true,
     },
+    viewSwitcherEntries: {
```
Comment (aab-odoo) [inline addons/web/static/src/search/with_search/with_search.js:109]: not sure about adding viewSwitcherEntries and breadcrumbs to the searchmodel.

File: addons/web/static/src/views/layout.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        this.components = { ...this.constructor.components, ...this.props.components };
+    }
+
+    get display() {
+        if (this.props.display) {
+            return this.props.display;
+        }
+        if (this.env.searchModel) {
+            return this.env.searchModel.display;
+        }
```
Comment (aab-odoo) [inline addons/web/static/src/views/layout.js:19]: Do we need these fallbacks? In general, I don't like when there are several way to do something.

Comment (aab-odoo) [reply addons/web/static/src/views/layout.js:19]: same for components in setup

File: addons/web/static/tests/views/layout_tests.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        const root = await makeWithSearch({
+            serverData,
+            Component: ToyComponent,
+            resModel: "foo",
+            searchViewId: false,
+        });
+
+        assert.hasClass(root.el, "o_toy_view o_view_sample_data");
+        assert.containsOnce(root, ".o_control_panel .o_cp_top_right .toy_search_bar");
+        assert.containsOnce(root, ".o_component_with_search_panel .o_search_panel");
```
Comment (aab-odoo) [inline addons/web/static/tests/views/layout_tests.js:70]: there is a searchpanel by default? If so it's consistent with the CP, but I'm not sure it's what we want

File: addons/web/static/tests/views/graph_view_tests.js [hunk 10]
```diff
@@ -2039,25 +2040,6 @@ QUnit.module("Views", (hooks) => {
         assert.containsNone(graph, ".abc");
     });
 
-    QUnit.test("no content helper after update", async function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/graph_view_tests.js:2042]: test removed?

File: addons/web/static/src/search/control_panel/control_panel.xml (hunk 6 above)
Comment (Arcasias) [reply addons/web/static/src/search/control_panel/control_panel.xml:44]: Yes, it's always set by the SearchModel

File: addons/web/static/src/search/with_search/with_search.js (hunk 7 above)
Comment (Arcasias) [reply addons/web/static/src/search/with_search/with_search.js:109]: It's to avoid passing props the the layout component, although @Polymorphe57 suggested we use a separate object in the `env` to contain all the view/action-specific information, to avoid overloading the SearchModel. I think this could be a discussion for an ulterior PR

File: addons/web/static/src/views/layout.js (hunk 8 above)
Comment (Arcasias) [reply addons/web/static/src/views/layout.js:19]: I think the `components` case makes sense because it falls back to the generic `ControlPanel` and `SearchPanel` classes, which we don't necessarily want to specifiy on each view class definition.

As for the `display` property it is purely theoretical: I wanted the `Layout` component to be independant of the View/SearchModel ecosystem, but still rely on it when it can. We can remove this if you want, as the layout will probably only be instanciated through views and actions.

File: addons/web/static/tests/views/graph_view_tests.js (hunk 10 above)
Comment (Arcasias) [reply addons/web/static/tests/views/graph_view_tests.js:2042]: Duplicate test. Might as well remove it in this PR ¯\\\_(ツ)\_/¯

File: addons/web/static/tests/views/layout_tests.js (hunk 9 above)
Comment (Arcasias) [reply addons/web/static/tests/views/layout_tests.js:70]: The search panel is displayed if all the following conditions are met:
- a `<searchpanel>` node is found in the search arch;
- the `display.searchPanel` prop has not explicitely been set to a falsy value;
- and either the `view.type` of the search model is included in the `view_types` attribute on the search panel node (default: list & kanban) OR there is no view type on the model, in which case the current component is considered an action and should always display its search panel unless explicitely told.

These toy component don't give any view information and are considered actions. I think we want a search panel in such cases.

File: addons/web/static/src/search/control_panel/control_panel.js [hunk 11]
```diff
@@ -21,6 +21,51 @@ export class ControlPanel extends Component {
...
+        // Clean up
+        Object.assign(slots, originalSlots);
+        return res;
+    }
+
+    /**
+     * @returns {Object[]}
+     */
+    get breadcrumbs() {
+        return this.props.breadcrumbs || this.env.searchModel.breadcrumbs;
```
Comment (aab-odoo) [inline addons/web/static/src/search/control_panel/control_panel.js:66]: do we need this double API? It would be nice if we could always look in the searchModel, or always in props

File: addons/web/static/tests/search/search_panel_tests.js [hunk 12]
```diff
@@ -286,6 +286,28 @@ QUnit.module("Search", (hooks) => {
...
+                    assert.step(`${method || route} on ${model}`);
+                }
+            },
+            Component: TestComponent,
+            resModel: "partner",
+            searchViewId: false,
+        });
+
+        assert.doesNotHaveClass(comp.el, "o_component_with_search_panel");
+        assert.deepEqual(getDomain(), []); // initial domain
```
Comment (aab-odoo) [inline addons/web/static/tests/search/search_panel_tests.js:308]: `assert.containsNone(comp, ".o_search_panel");`

File: addons/web/static/src/views/pivot/pivot_view.js [hunk 13]
```diff
@@ -129,8 +129,9 @@ export class PivotView extends Component {
         });
 
         // retrieve form and list view ids from the action
+        const views = this.env.actionContext.views || [];
```
Comment (aab-odoo) [inline addons/web/static/src/views/pivot/pivot_view.js:132]: I'd prefer actionContext not to be a requirement in the env (for instance, for hypothetical standalone views in the frontend)

Comment (aab-odoo) [reply addons/web/static/src/views/pivot/pivot_view.js:132]: or should we enforce it?

File: addons/web/static/src/search/with_search/with_search.js [hunk 14]
```diff
@@ -1,40 +1,50 @@
...
 export const SEARCH_KEYS = ["comparison", "context", "domain", "groupBy", "orderBy"];
 const OTHER_SEARCH_KEYS = ["irFilters", "searchViewArch", "searchViewFields", "searchViewId"];
 
 export class WithSearch extends Component {
     setup() {
         this.Component = this.props.Component;
 
+        const subEnv = {};
+        if (!this.env.actionContext) {
+            subEnv.actionContext = { id: false };
```
Comment (aab-odoo) [inline addons/web/static/src/search/with_search/with_search.js]: ok, we always have an actionContext :+1:

File: addons/web/static/src/search/with_search/with_search.js [hunk 15]
```diff
@@ -1,40 +1,50 @@
...
+            subEnv.__getContext__ = new CallbackRecorder();
         }
 
-        const SearchModelClass = this.Component.SearchModel || SearchModel;
-        this.searchModel = new SearchModelClass(this.env, {
+        useSubEnv(subEnv);
+
+        let SearchModelClass = SearchModel;
+        for (const key of this.env.actionContext.componentKeys || []) {
+            if (searchModelRegistry.contains(key)) {
```
Comment (aab-odoo) [inline addons/web/static/src/search/with_search/with_search.js:45]: weird? let's talk about this

File: addons/web/static/src/views/view.js [hunk 16]
```diff
@@ -10,14 +10,37 @@ import { useActionLinks } from "@web/views/helpers/view_hook";
 
 const viewRegistry = registry.category("views");
 
-const { Component } = owl;
+const { Component, hooks } = owl;
+const { useSubEnv } = hooks;
+
+export const makeActionContext = (...values) => {
```
Comment (aab-odoo) [inline addons/web/static/src/views/view.js]: this should maybe be moved to the action_service (like props, there is a common basis for both client and window actions, and then some keys only available for window actinos)

File: addons/crm/static/tests/forecast_view_tests.js [hunk 17]
```diff
@@ -135,7 +135,7 @@ QUnit.module("Views", (hooks) => {
 
             await makeView({
                 resModel: "foo",
-                type: "forecast_graph",
+                type: "graph",
```
Comment (aab-odoo) [inline addons/crm/static/tests/forecast_view_tests.js:138]: ? you don't test the forecast view then (?)

File: addons/web/static/src/webclient/actions/action_service.js [hunk 18]
```diff
@@ -10,7 +10,8 @@ import { KeepLast } from "@web/core/utils/concurrency";
 import { useBus } from "@web/core/utils/hooks";
 import { sprintf } from "@web/core/utils/strings";
 import { cleanDomFromBootstrap } from "@web/legacy/utils";
-import { makeActionContext, View } from "@web/views/view";
+import { makeActionContext } from "@web/search/with_search/with_search";
+import { View } from "@web/views/view";
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:14]: controller.context -> controller.actionContext (+ follow renaming)

File: addons/web/static/src/views/layout.js [hunk 19]
```diff
@@ -2,38 +2,23 @@
```
Comment (aab-odoo) [inline addons/web/static/src/views/layout.js:2]: should we move this file in another folder since it is not limited to views anymore?

**Review by aab-odoo — COMMENTED** [review]
I think the last solution is the best of all. We still have to find good names for actionContext, viewContext and layoutContext. For the last one, I suggest layoutComponents. I would like the other to stop using the word "context".

File: addons/web/static/src/views/layout.js [hunk 20]
```diff
@@ -2,38 +2,23 @@
...
-        if (!this.SearchPanel) {
-            this.SearchPanel = SearchPanel;
-        }
-    }
-
-    get display() {
-        return (this.env.searchModel && this.env.searchModel.display) || {};
+        this.components = makeLayoutContext(this.env.layoutContext);
+        this.display = (this.env.searchModel && this.env.searchModel.display) || {};
+        console.log(this.env);
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/views/layout.js:21]: Ouch

**Review by aab-odoo — APPROVED** [review]

File: addons/web/static/src/search/control_panel/control_panel.js [hunk 21]
```diff
@@ -97,6 +135,3 @@ ControlPanel.template = "web.ControlPanel";
 //         optional: true,
 //     },
 // };
-ControlPanel.defaultProps = {
```
Comment (aab-odoo) [inline addons/web/static/src/search/control_panel/control_panel.js:100]: we can remove the above comment as well

**Comment by aab-odoo** [comment]
robodoo rebase-merge r+
